includes/utils/class-amp-mutation-utils.php

```diff
@@ -0,0 +1,212 @@
+	 * @return array $histogram The incremented histogram.
+	 */
+	public static function increment_count( $histogram, $key ) {
+		$current_value     = isset( $histogram[ $key ] ) ? $histogram[ $key ] : 0;
+		$histogram[ $key ] = $current_value + 1;
+		return $histogram;
+	}
+
+	/**
+	 * Gets whether a node was removed in a sanitizer.
+	 *
+	 * @return boolean.
+	 */
+	public static function was_node_removed() {
+		return ! empty( self::$removed_nodes );
+	}
+
+	/**
+	 * Processes markup, to determine AMP validity.
+	 *
+	 * Passes $markup through the AMP sanitizers.
+	 * Also passes a 'mutation_callback' to keep track of stripped attributes and nodes.
+	 *
+	 * @param string $markup The markup to process.
+	 * @return void.
+	 */
+	public static function process_markup( $markup ) {
+		$args = array(
+			'content_max_width' => ! empty( $content_width ) ? $content_width : AMP_Post_Template::CONTENT_MAX_WIDTH,
+			'mutation_callback' => 'AMP_Mutation_Utils::track_removed',
+		);
+		AMP_Content_Sanitizer::sanitize( $markup, amp_get_content_sanitizers(), $args );
+	}
+
+	/**
+	 * Registers the REST API endpoint for validation.
+	 *
+	 * @return void.
+	 */
+	public static function amp_rest_validation() {
+		register_rest_route( 'amp-wp/v1', '/validate', array(
+			'methods'             => 'POST',
+			'callback'            => array( __CLASS__, 'validate_markup' ),
+			'args'                => array(
+				self::MARKUP_KEY => array(
+					'validate_callback' => array( __CLASS__, 'validate_arg' ),
+				),
+			),
+			'permission_callback' => array( __CLASS__, 'permission' ),
+		) );
+	}
+
+	/**
+	 * The permission callback for the REST request.
+	 *
+	 * @return boolean $has_permission Whether the current user has the permission.
+	 */
+	public static function permission() {
+		return current_user_can( 'edit_posts' );
+	}
+
+	/**
+	 * Validate the markup passed to the REST API.
+	 *
+	 * @param WP_REST_Request $request The REST request.
+	 * @return array|WP_Error.
+	 */
+	public static function validate_markup( WP_REST_Request $request ) {
+		$json = $request->get_json_params();
+		if ( empty( $json[ self::MARKUP_KEY ] ) ) {
+			return new WP_Error( 'no_markup', 'No markup passed to validator', array(
+				'status' => 404,
+			) );
+		}
+
+		return self::get_response( $json[ self::MARKUP_KEY ] );
+	}
+
+	/**
+	 * Gets the AMP validation response.
+	 *
+	 * If $markup isn't passed,
+	 * It will return the validation errors the sanitizers found in rendering the page.
+	 *
+	 * @param string $markup To validate for AMP compatibility (optional).
+	 * @return array $response The AMP validity of the markup.
+	 */
+	public static function get_response( $markup = null ) {
+		if ( isset( $markup ) ) {
+			self::process_markup( $markup );
+		}
+		$response = array(
+			'has_error'          => self::was_node_removed(),
+			'removed_nodes'      => self::$removed_nodes,
+			'removed_attributes' => self::$removed_attributes,
```

Comment:
It would be useful if the processed markup were also returned here. It could then be used for previewing, for example.

Reply:
Yes, that would help. Do you think we should return the processed markup only if we're validating a limited amount of markup, like a single Gutenberg block? We could detect this by whether `get_response()` is called with a `$markup` argument. 

```
if ( isset( $markup ) ) {
        $response['processed_markup] = $markup,
);
```

If `get_response()` is called without a `$markup` argument, it's validating the entire document.

Reply:
Thanks, @westonruter. 

[These commits](https://github.com/Automattic/amp-wp/compare/70521ef12c20139796108d09a588cf83de238486...0b6a4eef902ab76238ce7f758cbff9fb0b5b0474) output the `'processed_markup'`, when `$markup` is passed to the function.
<img width="973" alt="request-postman-validate" src="https://user-images.githubusercontent.com/4063887/35660340-139e2c06-06d1-11e8-9074-79c2bf7f0384.png">